File: python/src/train/layers/embedding.py
from __future__ import annotations

"""FeatureEmbeddings：离散特征索引 → 稠密嵌入拼接，支持变长序列池化。"""
import torch
import torch.nn as nn

FeatureTuple = tuple[str, int, int]
FeatureTensorMap = dict[str, torch.Tensor]


class FeatureEmbeddings(nn.Module):
    """特征嵌入层，支持 mean/sum/max pooling 处理变长序列特征。

    Args:
        features: [(name, vocab_size, embed_dim), ...]
        pooling_map: {name: "flatten"|"mean"|"sum"|"max"}
    """

    def __init__(
        self,
        features: list[FeatureTuple],
        pooling_map: dict[str, str] | None = None,
        total_dim: int | None = None,
    ) -> None:
        super().__init__()
        self.ordered_names = []
        self.feature_to_idx = {}
        self.pooling_map = pooling_map or {}
        total = 0
        for i, (name, vocab_size, embed_dim) in enumerate(features):
            self.feature_to_idx[name] = i
            self.ordered_names.append(name)
            setattr(self, f"emb_{name}", nn.Embedding(vocab_size, embed_dim))
            total += embed_dim
        self.num_features = len(features)
        self.total_dim = total_dim if total_dim is not None else total

    def _pool(self, emb: torch.Tensor, name: str) -> torch.Tensor:
        """Pool (batch, seq, dim) → (batch, pooled_dim)."""
        p = self.pooling_map.get(name, "first")
        if p == "mean":
            return emb.mean(dim=1)
        elif p == "sum":
            return emb.sum(dim=1)
        elif p == "max":
            return emb.max(dim=1).values
        elif p == "flatten":
            return emb.reshape(emb.shape[0], -1)  # (batch, seq, dim) → (batch, seq*dim)
        return emb[:, 0, :]

    def forward(self, x_inputs: FeatureTensorMap) -> torch.Tensor:
        """Lookup + pool + concat → [batch, total_dim]."""
        embeds = []
        for n in self.ordered_names:
            idx = x_inputs[n]  # (batch,) or (batch, seq)
            emb = getattr(self, f"emb_{n}")(idx)  # (batch, d) or (batch, seq, d)
            if idx.dim() == 2:
                emb = self._pool(emb, n)
            embeds.append(emb)
        return torch.cat(embeds, dim=1)

    def forward_stacked(self, x_inputs: FeatureTensorMap) -> list[torch.Tensor]:
        """Return list of [batch, 1, embed_dim] for FM interaction."""
        result = []
        for n in self.ordered_names:
            idx = x_inputs[n]
            emb = getattr(self, f"emb_{n}")(idx)
            if idx.dim() == 2:
                emb = self._pool(emb, n)
            result.append(emb.unsqueeze(1))
        return result

File: src/layers/embedding.rs
//! 特征嵌入层：FeatureSpec 定义 + FeatureEmbeddings 多特征合并。
use candle_core::{Result, Tensor};
use candle_nn::{embedding, Embedding, Module, VarBuilder};
use std::collections::HashMap;

use crate::feats::config::PoolingStrategy;
pub use crate::feats::FeatureSpec;

/// 多特征嵌入管理器：按序查找、池化、拼接。
pub struct FeatureEmbeddings {
    feature_to_idx: HashMap<String, usize>,
    ordered_names: Vec<String>,
    pooling: HashMap<String, PoolingStrategy>,
    embeddings: Vec<Embedding>,
    /// 特征数量。
    pub num_features: usize,
    /// 拼接后总输出维度。
    pub total_dim: usize,
}

impl FeatureEmbeddings {
    /// 从 FeatureSpec 列表构造嵌入层，每个特征创建独立 Embedding。
    pub fn new(vb: VarBuilder, features: &[FeatureSpec]) -> Result<Self> {
        let num_features = features.len();
        let mut feature_to_idx = HashMap::with_capacity(num_features);
        let mut ordered_names = Vec::with_capacity(num_features);
        let mut pooling = HashMap::with_capacity(num_features);
        let mut embeddings = Vec::with_capacity(num_features);
        let mut total_dim = 0;

        for (i, spec) in features.iter().enumerate() {
            feature_to_idx.insert(spec.name.clone(), i);
            ordered_names.push(spec.name.clone());
            pooling.insert(spec.name.clone(), spec.pooling);
            embeddings.push(embedding(
                spec.vocab_size,
                spec.embed_dim,
                vb.pp(format!("emb_{}", spec.name)),
            )?);
            total_dim += spec.output_dim();
        }

        Ok(Self {
            feature_to_idx,
            ordered_names,
            pooling,
            embeddings,
            num_features,
            total_dim,
        })
    }

    fn pool(&self, name: &str, emb: Tensor) -> Result<Tensor> {
        if emb.rank() != 3 {
            return Ok(emb);
        }

        match self.pooling.get(name).copied().unwrap_or_default() {
            PoolingStrategy::Mean => {
                let seq_len = emb.dim(1)? as f64;
                emb.sum(1)?.affine(1.0 / seq_len, 0.0)
            }
            PoolingStrategy::Sum => emb.sum(1),
            PoolingStrategy::Max => emb.max(1),
            PoolingStrategy::Flatten => {
                let batch = emb.dim(0)?;
                let seq_len = emb.dim(1)?;
                let dim = emb.dim(2)?;
                emb.reshape((batch, seq_len * dim))
            }
            PoolingStrategy::First => emb.narrow(1, 0, 1)?.squeeze(1),
        }
    }

    /// 前向：多特征嵌入 → 拼接为 [batch, total_dim] 张量。
    pub fn forward(&self, x_inputs: &HashMap<String, Tensor>) -> Result<Tensor> {
        let mut embeds = Vec::with_capacity(self.num_features);
        for name in &self.ordered_names {
            let input_tensor = x_inputs
                .get(name)
                .ok_or_else(|| candle_core::Error::Msg(format!("Feature '{}' not found", name)))?;
            let idx = *self.feature_to_idx.get(name).ok_or_else(|| {
                candle_core::Error::Msg(format!("Feature '{}' has no embedding index", name))
            })?;
            let emb = self.embeddings[idx].forward(input_tensor)?;
            embeds.push(self.pool(name, emb)?);
        }
        Tensor::cat(&embeds, 1)
    }

    /// 前向：返回各特征嵌入保留 seq 维度的 Vec。
    pub fn forward_stacked(&self, x_inputs: &HashMap<String, Tensor>) -> Result<Vec<Tensor>> {
        let mut embeds = Vec::with_capacity(self.num_features);
        for name in &self.ordered_names {
            let input_tensor = x_inputs
                .get(name)
                .ok_or_else(|| candle_core::Error::Msg(format!("Feature '{}' not found", name)))?;
            let idx = *self.feature_to_idx.get(name).ok_or_else(|| {
                candle_core::Error::Msg(format!("Feature '{}' has no embedding index", name))
            })?;
            let emb = self.embeddings[idx].forward(input_tensor)?;
            embeds.push(self.pool(name, emb)?.unsqueeze(1)?);
        }
        Ok(embeds)
    }
}
